opt_einsum/paths.py

```diff
@@ -652,6 +655,131 @@ def greedy(inputs, output, size_dict, memory_limit=None, choose_fn=None, cost_fn
+    #
+    # ((1,2),(0,(4,3))) --> [(1,2), (1,2), (0,2), (0,1)]
+    # 
+    # 0     0         0         (1,2)     --> ((1,2),(0,(3,4)))
+    # 1     3     --> (1,2) --> (0,(3,4))
+    # 2 --> 4         (3,4)
+    # 3     (1,2)
+    # 4
+    # 
+    # this function iterates through the table shown above from right to left; 
+    # note that contractions are commutative, i.e. (j,k) = (k,j);
+    # note that the solution is not unique;
+    
+    c = [c] # list of remaining contractions (lower part of columns shown above)
+    t = []  # list of elementary tensors (upper part of colums)
+    s = []  # resulting contraction sequence
+    
+    while len(c) > 0:
+        j, k = c.pop(-1)
+        s.insert(0, tuple())
+        
+        for i in sorted([i for i in [j, k] if type(i) == int]):
+            s[0] += (sum(1 for q in t if q < i),)
+            t.insert(s[0][-1], i)
+        
+        for i in [i for i in [j, k] if type(i) != int]:
+            s[0] += (len(t) + len(c),)
+            c.append(i)
+    
+    return s
+
+
+class DynamicProgrammingOptimizer(PathOptimizer):
+    
+    def __call__(self, inputs, output, size_dict, memory_limit=None):
+        
+        union = lambda a: functools.reduce(lambda x, y: x | y, a, set())
+        prod = lambda a: functools.reduce(lambda x, y: x * y, a, 1)
+        
+        # x[n_tensors][set of n_tensors tensors] = (indices, cost, contraction)
+        # use the immutable frozenset because set is not hashable
+        x = [
+            None, # x[0] is never used (just a placeholder)
+            {frozenset([j]): (i, 0, j) for j, i in enumerate(inputs)}
+        ]
+        
+        for n in range(2, len(inputs) + 1):
+            # compute x[n]
+            x.append(dict())
+            
+            # try to combine solutions from x[m] and x[n-m]
+            for m in range(1, n // 2 + 1):
+                for s1, (i1, cost1, cntrct1) in x[m].items():
+                    for s2, (i2, cost2, cntrct2) in x[n-m].items():
+                        if len(s1 & s2) == 0: # s1 and s2 have to be disjoint
+                        
+                            # avoid e.g. s1={0}, s2={1} and s1={1}, s2={0}
+                            if len(s1) != len(s2) or min(s1) < min(s2): 
+                            
+                                # ignore purely outer products: either tensor sets s1 and s2 are connected by a 
+                                # non-output index or none of them has remaining summation indices left
+                                if len((i1 & i2) - output) > 0 or (i1 & output == i1 and i2 & output == i2):
+                                    cost = cost1 + cost2 + prod(size_dict[j] for j in i1 | i2)
+                                    s = s1 | s2
+                                    if s not in x[n] or cost < x[n][s][1]:
+                                        r = set(range(len(inputs))) - (s1 | s2) # remaining set of tensors
```

Comment:
Real micro-optimization stuff, but there's a few computations you don't need to repeat here:
 * ``s1 | s2``
 * ``(i1 & i2) - output``
 * ``i1 | i2``

Reply:
You are right. I didn't really optimise this code up to now.